text/0082-entity-tags.md

```diff
@@ -0,0 +1,251 @@
+
+For entity types with no `tags` declaration in the schema, the validator gives `hasTag` the type `False` (to support short-circuiting), and rejects `getTag`.
+
+
+### Validating and parsing entities
+
+The Cedar authorizer's `is_authorized` function can be asked to validate that entities in a request are consistent with a provided schema. Extending validation to work with tags is straightforward. We check that every value in the `tags` map for an entity `e` has type `T` given that type of `e` is declared to have `tags T`.  If the type `e` doesn't have `tags` in the schema, then specifying a `tags` field is an error.
+
+Similarly, schema-based parsing considers schemas when parsing in entities, and it can confirm when parsing that the contents of `tags` attribute in the JSON have the appropriate shape, and if the type `T` has different behavior under schema-based parsing (for instance, if it is an extension type), then schema-based parsing can interpret the `tags` appropriately.
+
+
+### Symbolic compilation
+
+Tags as defined here have an efficient logical representation as *uninterpreted binary functions*.  In particular, for each declaration `entity E ... tags T`, we introduce an uninterpreted function `f_E` of type `E -> String -> Option T`.  With this representation, we can straightforwardly translate the `hasTag` and `getTag` operations as applications of the function `f_E`, where the absence of a key is denoted by binding it to the value `(none T)`.
+
+
+## Drawbacks
+
+Entity types are permitted to have a single `tags` declaration `tags T`, which eliminates the possibility of attaching multiple tags-maps to a single entity (unlike [RFC 68](https://github.com/cedar-policy/rfcs/pull/68)).  This RFC also does not support tags containing other tags-maps, comparing tags-maps with `==`, and the like (as discussed above).
+
+The use-cases that we are aware of do not suffer due to these limitations. If you wanted tags-maps to contain other tags-maps, or you wanted to store tags-maps in the `context` record, or you wanted to attach multiple tags-maps to a single entity, you can create specific entity types with those tags. For example:
+
+```
+entity User {
+  roles: Set<String>,
+  accessLevels: IntTags
+} tags StringTags;
+entity StringTags {
+} tags String;
+entity IntTags {
+} tags Long;
+```
+
+In effect, the `User`'s tags is equivalent to `tags (tags String)`, but we have added a level of indirection by expressing the inner `(tags String)` value as an entity with `tags String`. Similarly, the `User`'s `accessLevels` attribute is equivalent to `(tags Long)`. So, we can express policies such as:
+
+
+```
+permit(principal is User, action, resource is Document)
+when {
+  principal.roles.contains(context.role) &&
+  principal.accessLevels.hasTag(context.role) &&
+  principal.accessLevels.getTag(context.role) > 3
+};
+
+permit(principal is User, action, resource is Document)
+when {
```

Comment:
Is there any reason the two policies in this example could not be merged into one?

Reply:
Assuming that the policies are evaluated against inputs that conform to the example schema, we can re-write these two policies as follows:

```
permit(principal is User, action, resource is Document)
when {
  (principal.roles.contains(context.role) &&
   principal.accessLevels.hasTag(context.role) &&
   principal.accessLevels.getTag(context.role) > 3) || 
  (principal.hasTag("clearance") &&
   principal.getTag("clearance").hasTag("level") &&
   principal.getTag("clearance").getTag("level") == "top secret")
};
```

In general though, merging two policies like this is not a safe transformation (it can result in different authz results) because of errors, but that's a separate can of worms :)